### PR TITLE
add containerd http proxy configure daemonset

### PR DESCRIPTION
**Pull request by ycjiang50**
This daemonset allows users to configure containerd http proxy.

### CODE REVIEW

**Comment by vinayakankugoyal** [comment]
Can you also run shellcheck as part of presubmit?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+          echo "Reloading systemd management configuration"
+          systemctl daemon-reload
+          echo "Restarting containerd..."
+          systemctl restart containerd
+          END
+          )
+
+          CONFIG_PATH="/etc/systemd/system/containerd.service.d/http-proxy.conf"
+          CHECK_INTERVAL_SECONDS="30"
+          EXEC="nsenter -t 1 -m -u -i -n -p --"
```
Comment (vinayakankugoyal) [inline containerd-http-proxy/configure_http_proxy.yaml]: why is this required?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+          validate_proxy "${HTTP_PROXY:-}"
+          validate_proxy "${HTTPS_PROXY:-}"
+          validate_proxy "${NO_PROXY:-NO_PROXY=localhost}"
+
+          STARTUP_SCRIPT=$(cat <<END
+          set -e
+          set -u
+          mkdir -p /etc/systemd/system/containerd.service.d
+          echo "[Service]
```
Comment (vinayakankugoyal) [inline containerd-http-proxy/configure_http_proxy.yaml]: If this write is bad then does containerd not work, right?

File: containerd-http-proxy/sample_configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: containerd-proxy-configmap
+data:
+  HTTP_PROXY: HTTP_PROXY=http://proxy.example.com:80
+  HTTPS_PROXY: HTTPS_PROXY=http://proxy.example.com:443
```
Comment (vinayakankugoyal) [inline containerd-http-proxy/sample_configmap.yaml]: https?

**Comment by ycjiang50** [comment]
I'm not sure how to run shellcheck as github presubmit, but I ran it locally and have fixed two small issues.

**Comment by ycjiang50** [comment]
Hi @SergeyKanzhelev, could you take a look please?

File: containerd-http-proxy/sample_configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: containerd-proxy-configmap
+data:
+  HTTP_PROXY: HTTP_PROXY=http://proxy.example.com:80
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/sample_configmap.yaml]: why does it have to start with the `HTTP_PROXY=` and not simply contain the URL? Simple URL will make validator a bit easier

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          Environment=${HTTP_PROXY}
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml]: `${HTTP_PROXY}` must be in quotes

```suggestion
          Environment="${HTTP_PROXY}"
```

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        securityContext:
+          privileged: true
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -e
+          set -u
+
+          validate_proxy() {
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml:36]: what is the behavior if the url is empty? Will anything break?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            return 0
+          }
+
+          while true; do
+            $EXEC stat "${CONFIG_PATH}" > /dev/null 2>&1 && err=0 || err=$?
+            if [ "${err}" != 0 ]; then
+              do_startup_script
+            fi
+
+            sleep "${CHECK_INTERVAL_SECONDS}"
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml]: why do we need to check it continuously? Is there a chance that it will be deleted and we need to restore it back? Or this is just a protection from the failed configuration (failure in `do_startup_script`)?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+          validate_proxy "${NO_PROXY:-NO_PROXY=localhost}"
+
+          STARTUP_SCRIPT=$(cat <<END
+          set -e
+          set -u
+          mkdir -p /etc/systemd/system/containerd.service.d
+          echo "[Service]
+          Environment=${HTTP_PROXY}
+          Environment=${HTTPS_PROXY}
+          Environment=${NO_PROXY:-NO_PROXY=localhost}" > /etc/systemd/system/containerd.service.d/http-proxy.conf
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml]: Maybe variable `CONFIG_PATH` can be defined higher and reused here?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            fi
+
+            echo "!!! configure http-proxy succeeded!" 1>&2
+            return 0
+          }
+
+          while true; do
+            $EXEC stat "${CONFIG_PATH}" > /dev/null 2>&1 && err=0 || err=$?
+            if [ "${err}" != 0 ]; then
+              do_startup_script
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml]: with `set -e` on top, failure of this function will exit this loop. Is it a desired behavior?

Please note, the restartPolicy is Always on this container so this whole script will be restarted anyways. So this may be what is expected. I am just not sure

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 5 above)
Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml]: oh, just realized it is even worse than that. We need quotes for bash AND quotes written into the file itself.

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        - /bin/sh
+        - -c
+        - |
+          set -e
+          set -u
+
+          validate_proxy() {
+            input_string=$1
+
+            if echo "$input_string" | grep -qvE '^(HTTP_PROXY|HTTPS_PROXY|NO_PROXY)='; then
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml]: As this is implemented now, customer can set environment variable we do not intent. I didn't test it, but this doc: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html suggests that a single `Environment=` can be set to declare multiple variables. So the value of `HTTP_PROXY="http://proxy.example.com:80" ANOTHER_VAR="BAR"` will work

Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml]: hm, thought about it overnight. This may be handled by whitespace check. Not sure if one can combine a string that will look like multiple variables. Putting value in quotes will definitely help, but also check that value doesn't have quotes itself.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 7 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: It may not be necessary. But we need to keep the pod running otherwise the pod will be in CrashLoopBackOff status after script completes. I replaced it with `exec tail -f /dev/null` which will keep the script running.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 9 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: I see, so since the restartPolicy will restart the script on failures, I don't think this loop will be needed. Removed in new commit.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 8 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: The other use were removed.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 5 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: Added quote: Environment="HTTP_PROXY=${HTTP_PROXY}"

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 10 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: Added quote check in validation.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 6 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:36]: If the user doesn't specify HTTP_PROXY and HTTPS_PROXY in the configmap, the daemonset creation will fail because of missing env value. If they put empty string in configmap, validate_proxy() won't reject it.

Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml:36]: so if somebody configured an empty string - will everything work? I simply don't know how go will react on empty proxy

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+          validate_proxy "${HTTPS_PROXY:-}"
+          validate_proxy "${NO_PROXY:-localhost}"
+
+          STARTUP_SCRIPT=$(cat <<'END'
+          set -e
+          set -u
+          mkdir -p /etc/systemd/system/containerd.service.d
+          cat > /etc/systemd/system/containerd.service.d/http-proxy.conf <<EOF
+          [Service]
+          Environment="HTTP_PROXY=${HTTP_PROXY}"
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml:61]: just checking - this quote will be resuling in a quote in config file, correct? Bash is weird so wanted to doublecheck

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 6 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:36]: If users have empty string in configmap, the pod will run successfully and http-proxy.conf will have something like`Environment="HTTP_PROXY="`. They will get image pull error if they want to pull new images because proxy is invalid. I think it'll be the same if they have any typos in proxy, but they can always fix the configmap and redeploy the daemonset.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 11 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:61]: Correct. I verified there'll be quotes in config file.

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      labels:
+        name: containerd-http-proxy
+    spec:
+      hostPID: true
+      containers:
+      - name: startup-script
+        image: gke.gcr.io/debian-base:bookworm-v1.0.0-gke.1
+        imagePullPolicy: IfNotPresent
+        securityContext:
+          privileged: true
```
Comment (vinayakankugoyal) [inline containerd-http-proxy/configure_http_proxy.yaml:25]: why is this privileged?

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 1 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: Running systemctl or restart containerd would need it. Mounting all the required directories and files is impractical.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 2 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: It should only impact proxy, like unable to pull images if the proxy urls are invalid.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: It'll get permission denied error on some system files without privileged:true.

File: containerd-http-proxy/sample_configmap.yaml (hunk 3 above)
Comment (ycjiang50) [reply containerd-http-proxy/sample_configmap.yaml]: Updated.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (vinayakankugoyal) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: Can you share what the error was?

Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: It got `nsenter: cannot open /proc/1/ns/ipc: Permission denied`

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+          STARTUP_SCRIPT=$(cat <<'END'
+          set -e
+          set -u
+          mkdir -p /etc/systemd/system/containerd.service.d
+          cat > /etc/systemd/system/containerd.service.d/http-proxy.conf <<EOF
+          [Service]
+          Environment="HTTP_PROXY=${HTTP_PROXY}"
+          Environment="HTTPS_PROXY=${HTTPS_PROXY}"
+          Environment="NO_PROXY=${NO_PROXY:-localhost}"
```
Comment (tallclair) [inline containerd-http-proxy/configure_http_proxy.yaml:63]: Consider escaping these for added safety:
```suggestion
          $(printf 'Environment="HTTP_PROXY=%q"' "${HTTP_PROXY}")
          $(printf 'Environment="HTTPS_PROXY=%q"' "${HTTPS_PROXY}")
          $(printf 'Environment="NO_PROXY=%q"' "${NO_PROXY:-localhost}")
```

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+              echo "Error: Input cannot contain quotes. Input: '$input_string'"
+              exit 1
+            fi
+          }
+
+          validate_proxy "${HTTP_PROXY:-}"
+          validate_proxy "${HTTPS_PROXY:-}"
+          validate_proxy "${NO_PROXY:-localhost}"
+
+          STARTUP_SCRIPT=$(cat <<'END'
```
Comment (tallclair) [inline containerd-http-proxy/configure_http_proxy.yaml]: Rather than the multiple layers of indirection here, it seems like it would be a lot simpler to just mount `/etc/systemd/system/containerd.service.d` as a host volume and write the script directly from here. Is there a reason you can't use a host mount?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+          systemctl daemon-reload
+          echo "Restarting containerd..."
+          systemctl restart containerd
+          END
+          )
+
+          EXEC="nsenter -t 1 -m -u -i -n -p --"
+          $EXEC sh -c "${STARTUP_SCRIPT}"
+          echo "!!! configure http-proxy succeeded!" 1>&2
+          exec tail -f /dev/null
```
Comment (tallclair) [inline containerd-http-proxy/configure_http_proxy.yaml]: I think it would be more robust and lower footprint to execute this script in an init container, and then just use the pause image as the main "workload".

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (tallclair) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: You might need `hostIPC: true`

Comment (vinayakankugoyal) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: Is hostIPC needed, it seems like we are entering ipc namespace even though that is not required. Could we  scope the nsenter to just pid (and user if required).

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 13 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:63]: %q seems not supported by dash. We had to use dash because bash is not included in the image.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 15 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: Thanks, updated.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 14 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: That's good idea. Updated.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: Yes, I removed the unnecessary namespaces.

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 16]
```diff
@@ -42,52 +45,31 @@ spec:
...
+          [Service]
+          Environment="HTTP_PROXY=${HTTP_PROXY}"
+          Environment="HTTPS_PROXY=${HTTPS_PROXY}"
+          Environment="NO_PROXY=${NO_PROXY:-localhost}"
+          EOF
+
+          EXEC="nsenter -t 1 -m -p --"
           echo "Reloading systemd management configuration"
-          systemctl daemon-reload
+          $EXEC sh -c "systemctl daemon-reload"
```
Comment (tallclair) [inline containerd-http-proxy/configure_http_proxy.yaml]: nit: Do you need `sh -c` here? Can you just nsenter to execute `systemctl` directly?

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (vinayakankugoyal) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: I believe you can remove privileged then and add securityContext.capabilities.add [SYS_ADMIN]?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+          mountPath: /etc/systemd/system/containerd.service.d
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -e
+          set -u
+
+          validate_proxy() {
+            input_string=$1
```
Comment (vinayakankugoyal) [inline containerd-http-proxy/configure_http_proxy.yaml:37]: what kind of values can the input string be?
can we add a check that makes sure the match the regex of expected values?

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: It got "nsenter: cannot open /proc/1/ns/pid: Permission denied" without privilege

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 16 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: Right, we don't need sh -c here. Removed.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 12 above)
Comment (vinayakankugoyal) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: when you removed privileged did you add SYS_ADMIN to the capabilities of the container?

Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:25]: Yes, I did

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 17 above)
Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:37]: They should be urls. NO_PROXY could have multiple address separated by comma, and allow wildcard. Dash doesn't support using regex very well and I'm worried about scenarios where some valid cases could be rejected. And I think checking space and newline should be sufficient.

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 1 above)
Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml]: btw, this is what we already have in this repo: https://github.com/GoogleCloudPlatform/k8s-node-tools/blob/6f5f17af0971c54ddff77b000abda3788cba0965/container-insecure-registry/insecure-registry-config.yaml#L46

Would using the same image be better instead of nsenter?

Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml]: In other examples we do this: https://github.com/GoogleCloudPlatform/k8s-node-tools/blob/6f5f17af0971c54ddff77b000abda3788cba0965/configure-hugepages/configure-hugepages.yaml#L103

Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml]: not sure what chroot /host does and why we don't do it here

Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml]: We once had the discussion with @vinayakankugoyal but the gcr.io/google-containers/startup-script:v1 image is old and unmaintained.

Comment (wangzhen127) [reply containerd-http-proxy/configure_http_proxy.yaml]: The [implementation](https://github.com/kubernetes-retired/contrib/blob/master/startup-script/manage-startup-script.sh) of gcr.io/google-containers/startup-script is 8 years old, but still looks ok. @SergeyKanzhelev Is there any recommended way? Or it is case by case?

**Review by tallclair — COMMENTED** [review]
lgtm

**Review by vinayakankugoyal — APPROVED** [review]

File: containerd-http-proxy/configure_http_proxy.yaml (hunk 1 above)
Comment (SergeyKanzhelev) [reply containerd-http-proxy/configure_http_proxy.yaml]: do we need chroot /host for nsenter? Do you know why it is used in other Daemonsets in this repo?

File: containerd-http-proxy/configure_http_proxy.yaml [hunk 18]
```diff
@@ -0,0 +1,94 @@
+kind: DaemonSet
```
Comment (SergeyKanzhelev) [inline containerd-http-proxy/configure_http_proxy.yaml:1]: is there anything we need to add to whitelist this? Should we add a note here for the future reviews that even minor modifications of this yaml will result in the broken whitelist in Autopilot?

Comment (ycjiang50) [reply containerd-http-proxy/configure_http_proxy.yaml:1]: Added a README with instructions as well.

**Review by SergeyKanzhelev — APPROVED** [review]
